Description: Use GnuTLS instead of OpenSSL
Author: Joey Hess <joeyh@debian.org>, patch creation: Arno Töll <debian@toell.net>
Last-Update: 2011-06-30

* snowdrop.c: Add GnuTLS header
* lang-eng.c: Add GnuTLS header
* Makefile:
  + Make the Makefile honor $(DESTDIR)
  + Add GnuTLS build time dependencies
  + Don't hide CFLAGS from dpkg-buildflags if used
    [Arno Töll]

--- a/snowdrop.c
+++ b/snowdrop.c
@@ -28,6 +28,9 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <time.h>
+#ifdef USE_GNUTLS
+#include <gnutls/openssl.h>
+#else
 #ifdef USE_OPENSSL
 #include <openssl/md5.h>
 #else
@@ -37,6 +40,7 @@
 #define MD5_Final  MD5Final
 #define MD5_Update MD5Update
 #endif /* USE_OPENSSL */
+#endif /* USE_GNUTLS */
 
 #include <string.h>
 
--- a/Makefile
+++ b/Makefile
@@ -26,9 +26,10 @@
 LANG    = eng engf c
 
 BINROOT = /usr/bin/
+DESTDIR = /
 
-CFLAGS  = -ggdb -O9 -fomit-frame-pointer -funroll-loops -fexpensive-optimizations \
-          -ffast-math -Wall 
+CFLAGS  += -ggdb -O2 -fomit-frame-pointer -funroll-loops -fexpensive-optimizations \
+          -ffast-math -Wall -g
 
 
 all: modules snowdrop toinstall
@@ -36,26 +37,26 @@
 modules: language.h
 	@echo ; \
 	echo "[*] Compiling language modules:" ; \
-	test -d /usr/include/openssl && USEOPENSSL=1; \
-	test -d /usr/local/include/openssl && USEOPENSSL=1; \
-	test "$$USEOPENSSL" = "" || echo "[+] Using OpenSSL MD5 modules." ; \
+	test -f /usr/include/gnutls/openssl.h && USEOPENSSL=1; \
+	test -f /usr/local/include/gnutls/openssl.h && USEOPENSSL=1; \
+	test "$$USEOPENSSL" = "" || echo "[+] Using GNU TLS MD5 modules." ; \
 	test "$$USEOPENSSL" = "" && echo "[+] Trying to use RSA MD5 modules." ; \
 	for i in $(LANG); do \
 	  echo "[+] Building language module for '$$i'..."; \
-	  test "$$USEOPENSSL" = "" || ADDME="-DUSE_OPENSSL" ; \
+	  test "$$USEOPENSSL" = "" || ADDME="-DUSE_OPENSSL -DUSE_GNUTLS" ; \
 	  $(CC) $$ADDME $(CFLAGS) -c lang-$$i.c -o lang-$$i.o || exit 1; \
 	done; \
 	echo "[*] Language modules compiled."
 
 snowdrop: snowdrop.c language.h 
 	@echo "[*] Compiling main code:"; \
-	test -d /usr/include/openssl && USEOPENSSL=1; \
-	test -d /usr/local/include/openssl && USEOPENSSL=1; \
+	test -f /usr/include/gnutls/openssl.h && USEOPENSSL=1; \
+	test -f /usr/local/include/gnutls/openssl.h && USEOPENSSL=1; \
 	for i in $(LANG); do \
 	  echo "[+] Building 'sd-$$i'..." ; \
 	  ADDME="-lmd5"; \
-	  test "$$USEOPENSSL" = "" || ADDME="-DUSE_OPENSSL -lcrypto" ; \
-	  $(CC)  -DVER=\"$(VER)\" $(CFLAGS) -DTARGETLANG=\"$$i\" snowdrop.c lang-$$i.o -o sd-$$i $$ADDME || exit 1; \
+	  test "$$USEOPENSSL" = "" || ADDME="-DUSE_OPENSSL -DUSE_GNUTLS -lgnutls-extra -lgnutls-openssl $(shell pkg-config --libs --cflags gnutls)" ; \
+	  $(CC) -DVER=\"$(VER)\" $(CFLAGS) -DTARGETLANG=\"$$i\" snowdrop.c lang-$$i.o -o sd-$$i $$ADDME || exit 1; \
 	done; \
 	echo "[*] Main code compiled."
 
@@ -67,12 +68,12 @@
 clean:
 	rm -f sd-* *.o core core.* a.out
 
-install: modules snowdrop
-	@echo "[*] Installing binaries in $(BINROOT)..."
-	cp -f sd-* $(BINROOT)
+install:
+	@echo "[*] Installing binaries in $(DESTDIR)$(BINROOT)..."
+	cp -f sd-* $(DESTDIR)$(BINROOT)
 	@echo "[*] Installing synonyms database..."
-	@mkdir /usr/share/snowdrop || true
-	cp synonyms /usr/share/snowdrop/
+	@mkdir $(DESTDIR)/usr/share/snowdrop || true
+	cp synonyms $(DESTDIR)/usr/share/snowdrop/
 	@echo "[*] Installation complete."
 
 publish: clean
--- a/lang-eng.c
+++ b/lang-eng.c
@@ -24,6 +24,9 @@
 #include <assert.h>
 #include <string.h>
 
+#ifdef USE_GNUTLS
+#include <gnutls/openssl.h>
+#else
 #ifdef USE_OPENSSL
 #include <openssl/md5.h>
 #else
@@ -33,6 +36,7 @@
 #define MD5_Final  MD5Final
 #define MD5_Update MD5Update
 #endif /* USE_OPENSSL */
+#endif /* USE_GNUTLS */
 
 #include "language.h"
 
